lgn-provers/src/params/mod.rs

```diff
@@ -1,160 +1,247 @@
+                    },
+                    None => {},
+                }
+            },
+        };
+    }
 
-    Ok(bytes)
+    bail!(
+        "Download failed after retries. filepath: {} retries: {}",
+        filepath.display(),
+        DOWNLOAD_BACKOFF_RETRIES,
+    );
 }
 
-/// Download the content from `file_name` under `base_url`, ensuring that its checksum matches
-/// the provided `expected_checksum`.
-fn download_file(
+async fn resume_download(
+    buf: &mut Vec<u8>,
     base_url: &str,
-    file_name: &str,
+    file: &mut File,
+    filepath: &Path,
+    fileurl: &str,
+    retry: usize,
+    client: &reqwest::Client,
+    hasher: &mut blake3::Hasher,
     expected_checksum: &blake3::Hash,
-) -> anyhow::Result<Bytes> {
-    let file_url = format!("{base_url}/{file_name}");
-    info!("downloading params from {}", file_url);
-
-    let client = reqwest::blocking::Client::builder()
-        .timeout(std::time::Duration::from_secs(HTTP_TIMEOUT))
-        .build()
-        .context("building reqwest client")?;
+) -> anyhow::Result<()> {
+    let metadata = file.metadata().await?;
+
+    info!(
+        "Downloading params. base_url: {} filepath: {} size: {}MiB retry: {}",
+        base_url,
+        filepath.display(),
+        metadata.len() / BYTES_TO_MEGABYTES_U64,
+        retry,
+    );
 
     let response = client
-        .get(file_url)
+        .get(fileurl)
+        .header("Range", format!("bytes={}-", metadata.len()))
         .send()
-        .context("downloading params from remote")?;
+        .await?;
 
-    if !response.status().is_success() {
-        bail!(
-            "downloading params from remote: status = {}",
-            response.status()
+    if response.status() == StatusCode::RANGE_NOT_SATISFIABLE {
+        warn!(
+            "Local file is bigger than remote, resetting length and checking checksum. filepath: {}",
+            filepath.display(),
+        );
+        let response = client.head(fileurl).send().await?;
+        let length = response.headers().get("Content-Length").map_or(
+            Ok(0u64),
+            |v| -> anyhow::Result<u64> {
+                let as_ascii = v.to_str()?;
+                let parsed = u64::from_str(as_ascii)?;
+                Ok(parsed)
+            },
+        )?;
+
+        hasher.reset();
+        buf.resize(
+            length.try_into().expect("File size should fit in a usize"),
+            0,
         );
+        file.set_len(length).await?;
+        hasher.update_rayon(&buf);
+    } else {
+        ensure!(
+            response.status().is_success(),
+            "Requesting params failed. status: {} filepath: {}",
+            response.status(),
+            filepath.display(),
+        );
+
+        let mut stream = response.bytes_stream();
+        while let Some(data) = stream.next().await {
+            let data = data?;
+            file.write_all(&data).await?;
+            hasher.update_rayon(&data);
+            buf.extend(data);
+        }
```

Comment:
A few notes:

- This does two copies:
   - `kernel -> use_space` when the data arrives
   - `user_space -> kernel -> file` when appending to the file
- This can eventually be changed to `kernel -> file` and `kernel -> user_space`, maybe it is has better performance for these large files

This also keeps a copy of the data in-memory, in the `buf`, ideally that should go away or maybe be mmaped to save memory.

Reply:
Frankly I think that mmapping at download-time does not make sense while we need to load in-RAM for plonky2 in any case.